llama_index/llms/cohere.py

```diff
@@ -0,0 +1,306 @@
+    messages_to_cohere_history,
+)
+
+
+class Cohere(LLM):
+    model: str = Field(description="The cohere model to use.")
+    temperature: float = Field(description="The temperature to use for sampling.")
+    max_retries: int = Field(
+        default=10, description="The maximum number of API retries."
+    )
+    additional_kwargs: Dict[str, Any] = Field(
+        default_factory=dict, description="Additonal kwargs for the anthropic API."
+    )
+
+    _client: Any = PrivateAttr()
+    _aclient: Any = PrivateAttr()
+
+    def __init__(
+        self,
+        model: str = "command",
+        temperature: float = 0.5,
+        max_tokens: int = 512,
+        timeout: Optional[float] = None,
+        max_retries: int = 10,
+        api_key: Optional[str] = None,
+        additional_kwargs: Optional[Dict[str, Any]] = None,
+        callback_manager: Optional[CallbackManager] = None,
+    ) -> None:
+        try:
+            import cohere
+        except ImportError as e:
+            raise ImportError(
+                "You must install the `cohere` package to use Cohere."
+                "Please `pip install cohere`"
+            ) from e
+        additional_kwargs = additional_kwargs or {}
+        callback_manager = callback_manager or CallbackManager([])
+
+        self._client = cohere.Client(api_key)
+        self._aclient = cohere.AsyncClient(api_key)
+
+        super().__init__(
+            temperature=temperature,
+            additional_kwargs=additional_kwargs,
+            timeout=timeout,
+            max_retries=max_retries,
+            model=model,
+            callback_manager=callback_manager,
+        )
+
+    @classmethod
+    def class_name(cls) -> str:
+        """Get class name."""
+        return "Cohere_LLM"
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        return LLMMetadata(
+            context_window=cohere_modelname_to_contextsize(self.model),
+            num_output=self.max_tokens,
+            is_chat_model=True,
+            model_name=self.model,
+        )
+
+    @property
+    def _model_kwargs(self) -> Dict[str, Any]:
+        base_kwargs = {
+            "model": self.model,
+            "temperature": self.temperature,
+        }
+        return {
+            **base_kwargs,
+            **self.additional_kwargs,
+        }
+
+    def _get_all_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
+        return {
+            **self._model_kwargs,
+            **kwargs,
+        }
+
+    @llm_chat_callback()
+    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
+        history = messages_to_cohere_history(messages[:-1])
+        prompt = messages[-1].content
+        all_kwargs = self._get_all_kwargs(**kwargs)
+        if all_kwargs["model"] not in CHAT_MODELS:
+            raise ValueError(f"{all_kwargs['model']} not supported for chat")
+
+        if "stream" in all_kwargs:
+            all_kwargs["stream"] = False  # use stream_chat instead
```

Comment:
nit: helpful to raise a warning to nudge user to call `stream_chat` endpoint instead.

Reply:
Would it be better if I raise a warning and set stream to False, or just raise the warning and allow setting stream to True?